Update AddCoolingProfile.py

GregValiant opened the pull request:
Add control for a Build Volume fan.
Add 'Idle speed' for the inactive extruder rather than always going to "0".

## Type of change

- [ X] New features (non-breaking change which adds functionality)

**Test Configuration**:
* Operating System:
Win 10 Pro
All Cura versions
# Checklist:
<!-- Check if relevant -->

- [ X] My code follows the style guidelines of this project as described in [UltiMaker Meta](https://github.com/Ultimaker/Meta) and [Cura QML best practices](https://github.com/Ultimaker/Cura/wiki/QML-Best-Practices)
- [X ] I have read the [Contribution guide](https://github.com/Ultimaker/Cura/blob/main/CONTRIBUTING.md) 
- [ X] I have commented my code, particularly in hard-to-understand areas
- [ X] I have uploaded any files required to test this change

HellAholic commented:
Code could use a bit more cleaning up imo, some of the if statements for example could be simplified. You could also do the init() and define self.curaApp and use that instead of `Application.getInstance().getGlobalContainerStack()` or setting the curaApp in different functions or in the if statements.

GregValiant commented:
Changes made.
I also added the "if ';POSTPROCESSED' in data[0]" exit line so the script can't run twice.

A little history...This was the first script I'd written.  I took it on because I needed it for PETG, and because it had been a feature request here since 2017 and no one else had bothered.  I'm quite pleased that there has never been a bug report regarding it (knock-on-wood).

HellAholic commented on plugins/PostProcessingPlugin/scripts/AddCoolingProfile.py:
```diff
@@ -273,46 +280,141 @@ def getSettingDataString(self):
     def initialize(self) -> None:
         super().initialize()
-        scripts = Application.getInstance().getGlobalContainerStack().getMetaDataEntry("post_processing_scripts")
+        global_stack = Application.getInstance().getGlobalContainerStack()
+        extruder = global_stack.extruderList
+        extruder_count = global_stack.getProperty("machine_extruder_count", "value")
```
Since you're using these in the execute as well, it would be better to define as `self.` in the initialize and then reuse there in execute.
```suggestion
        self.global_stack = Application.getInstance().getGlobalContainerStack()
        self.extruder_list = self.global_stack.extruderList
        self.extruder_count = self.global_stack.getProperty("machine_extruder_count", "value")
```

GregValiant replied:
Completed

HellAholic commented on plugins/PostProcessingPlugin/scripts/AddCoolingProfile.py:
```diff
@@ -513,41 +624,44 @@ def execute(self, data):
+            return self._multi_fan_by_layer(data, layer_0_index, fan_list, t0_fan, t1_fan, t2_fan, t3_fan, fan_mode, off_fan_speed)
 
-        #Single Fan "By Feature"------------------------------------------
+        #Single Fan "By Feature"
```
:P
```suggestion
        # Single Fan "By Feature"
```

HellAholic commented on plugins/PostProcessingPlugin/scripts/AddCoolingProfile.py:
```diff
@@ -513,41 +624,44 @@ def execute(self, data):
             return self._single_fan_by_feature(data, layer_0_index, the_start_layer, the_end_layer, the_end_is_enabled, fan_list, t0_fan, feature_speed_list, feature_name_list, feature_fan_combing)
 
-        #Multi Fan "By Feature"-------------------------------------------
+        #Multi Fan "By Feature"
```
```suggestion
        # Multi Fan "By Feature"
```

HellAholic commented on plugins/PostProcessingPlugin/scripts/AddCoolingProfile.py:
```diff
@@ -834,4 +953,35 @@ def _add_travel_comment(self, comment_data: str, lay_0_index: str) -> str:
+            bv_data[len(bv_data)-2] += f"M106 S0 P{bv_fan_nr}\n"
+        else:
+            for index, layer in enumerate(bv_data):
+                if ";LAYER:" + str(bv_end_layer) + "\n" in layer:
```
```suggestion
                if f";LAYER:{bv_end_layer}\n" in layer:
```

GregValiant replied:
Completed

HellAholic commented on plugins/PostProcessingPlugin/scripts/AddCoolingProfile.py:
```diff
@@ -834,4 +953,35 @@ def _add_travel_comment(self, comment_data: str, lay_0_index: str) -> str:
+        if bv_end_layer != -1:
+            bv_end_layer -= 1
+        # Get the PWM speed or if RepRap then the 0-1 speed
+        if self.global_stack.extruderList[0].getProperty("machine_scale_fan_speed_zero_to_one", "value"):
```
```suggestion
        if self.extruder_list[0].getProperty("machine_scale_fan_speed_zero_to_one", "value"):
```

HellAholic replied:
This is in case you assign the current `extruder` as `self.extruder_list` during the initialize()

GregValiant replied:
Completed

GregValiant replied:
Done.

HellAholic commented on plugins/PostProcessingPlugin/scripts/AddCoolingProfile.py:
```diff
@@ -273,46 +280,141 @@ def getSettingDataString(self):
 
-        #Get some information from Cura-----------------------------------
-        extruder = mycura.extruderList
+        #Get some information from Cura
```
```suggestion
        # Get some information from Cura
```

HellAholic commented on plugins/PostProcessingPlugin/scripts/AddCoolingProfile.py:
```diff
@@ -273,46 +280,141 @@ def getSettingDataString(self):
+        print_sequence = str(self.global_stack.getProperty("print_sequence", "value"))
 
-        #Assign the fan numbers to the tools------------------------------
+        #Assign the fan numbers to the tools
```
```suggestion
        # Assign the fan numbers to the tools
```

HellAholic commented on plugins/PostProcessingPlugin/scripts/AddCoolingProfile.py:
```diff
@@ -324,16 +426,16 @@ def execute(self, data):
             if extruder_count > 3: t3_fan = " P" + str((extruder[3].getProperty("machine_extruder_cooling_fan_number", "value")))
 
-        #Initialize the fan_list with defaults----------------------------
+        #Initialize the fan_list with defaults
```
```suggestion
        # Initialize the fan_list with defaults
```

HellAholic commented on plugins/PostProcessingPlugin/scripts/AddCoolingProfile.py:
```diff
@@ -324,16 +426,16 @@ def execute(self, data):
             fan_list[num + 1] = "M106 S0"
 
-        #Assign the variable values if "By Layer"-------------------------
+        #Assign the variable values if "By Layer"
```
```suggestion
        # Assign the variable values if "By Layer"
```

HellAholic commented on plugins/PostProcessingPlugin/scripts/AddCoolingProfile.py:
```diff
@@ -729,27 +846,29 @@ def _multi_fan_by_feature(self, data: str, layer_0_index: int, the_start_layer:
+            multi_fan_data[-1] += "M106 S0 P1\nM106 S0 P0\n"
         return multi_fan_data
 
     #Try to catch layer input errors, set the minimum speed to 12%, and put the strings together
```
```suggestion
    # Try to catch layer input errors, set the minimum speed to 12%, and put the strings together
```

HellAholic commented on plugins/PostProcessingPlugin/scripts/AddCoolingProfile.py:
```diff
@@ -273,46 +280,141 @@ def getSettingDataString(self):
+        # No P parameter if there is a single fan circuit
                 t0_fan = ""
 
         #Get the cooling fan numbers for each extruder if the printer has multiple extruders
```
```suggestion
        # Get the cooling fan numbers for each extruder if the printer has multiple extruders
```

HellAholic commented on plugins/PostProcessingPlugin/scripts/AddCoolingProfile.py:
```diff
@@ -273,46 +280,141 @@ def getSettingDataString(self):
+        extruder = self.global_stack.extruderList
+        extruder_count = self.global_stack.getProperty("machine_extruder_count", "value")
 
         #This will be true when fan scale is 0-255pwm and false when it's RepRap 0-1 (Cura 5.x)
```
```suggestion
        # This will be true when fan scale is 0-255pwm and false when it's RepRap 0-1 (Cura 5.x)
```

HellAholic commented:
Some minor comments, looks good overall. Added some space after the # in a commit.

GregValiant commented:
The last commit should do the job.  All of your suggestions were implemented.

HellAholic commented:
Mainly looked at the changes introduced by the PR, but going through the entire code with a finer comb:
- Take a look at the try/except clauses -> Only should catch the specific exceptions you expect (e.g, ValueError, AttributeError,...) depending on the `try` content.
- Parts of the code such as the functions (eg. _multi_fan_by_feature) that handle multi extruder setups and fan control could be refactored to reduce the repetitive code -> this helps with future development and debugging as you would only change the logic of the helper functions instead of applying it to the main function. It lowers the complexity of the code and makes it also easier to follow and maintain.
- There are unused variables in the function definitions. Some examples '_multi_fan_by_layer: fan_mode', '_single_fan_by_feature: fan_list, index = 1', '_multi_fan_by_feature: layer_0_index, the_end_is_enabled, fan_list, fan_mode'
- Functions could use a bit of documentation, referencing the comment under the function that explains in a wide scope what the function does and what the parameters are and what the function returns:
```
def _single_fan_by_layer(self, data: str, layer_0_index: int, fan_list: str, t0_fan: str) -> str:
    """
    Adjusts fan speeds for a single fan by layer.

    :param data: The G-code data as a list of strings.
    :param layer_0_index: The index of the first layer in the G-code.
    :param fan_list: A list of layer/fan speed pairs.
    :param t0_fan: The fan identifier for the first extruder.
    :return: The modified G-code data.
    """
```
- Optional: You could use the logger to indicate when a function starts, helps with troubleshooting later and tracking the code execution during final testing.

GregValiant commented:
We run into my ignorance here.  I have no idea what an "attribute error" is.

At one point Remco asked that I try to simplify sections of the code.  That's when I split off the four separate 'main' functions.  Yes, I may have missed variables that became obsolete during those changes.

I'm not in the mood for going back over it again but I will at some point.  It's a bastard to debug and right now it works fine even though the code remains pretty ugly.

I've been working on ChangeAtZ and PauseAtHeight.  I figured I'd submit them after 5.10 stable is out.  Once I get that far I'll return to this script and work on it some more.

HellAholic commented:
I went over it since you made the request, take your time and see what you think could be added. It's best to work on things when you're motivated.
Probably you know "attribute error" but just not by the name. Think of it this way, what sort of error message shows up in the logs (or when you run the script) if you don't include the try/except, that's the type of exception you should handle in the except part.

Some General rambling to clarify.
examples for attribute error
```
class Car:
    def __init__(self, brand):
        self.brand = brand

my_car = Car("Toyota")
print(my_car.brannd)  # Typo: 'brannd' instead of 'brand'
```
AttributeError: 'Car' object has no attribute 'brannd'

```
class Person:
    def __init__(self, name):
        self.name = name

p = Person("Alice")
print(p.age)  # 'age' was never defined in the class
```
AttributeError: 'Person' object has no attribute 'age'

```
class Animal:
    def speak(self):
        return "Some sound"

dog = Animal()
print(dog.run())  # 'run()' is not defined in the class
```
AttributeError: 'Animal' object has no attribute 'run'

```
num = 10
print(num.append(5))  # Integers don't have an 'append' method
```
AttributeError: 'int' object has no attribute 'append'

GregValiant commented:
Another effort.

GregValiant commented:
I had changed some of the "except" lines to indicate the error to trap.  I was wrong in regards to line 782.  It had looked like it would be an IndexError (and it still might throw one) but it kicked out on a ValueError.  I have reverted back to the simple "except:" which did work.

GregValiant commented:
Changed an exception from IndexError to a ValueError.
